Add scripts and ITs for SAF (Simple Archive Format) import and export functionalities

francescopioscognamiglio opened the pull request:
## References
Fixes #2850
Fixes #7980
Fixes #2843

Frontend PR: https://github.com/DSpace/dspace-angular/pull/1827

## Description
Added new script configurations for the import and export via SAF.
Added ITs for the import and export via SAF.
Small improvements on SAF import functionality (e.g. in case of relationships via `relationships` or `metadata_relation.xml` file).

## Instructions for Reviewers
List of changes in this PR:
* SAF import and export functionalities as processes:
   * Modify dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportCLITool.java and dspace-api/src/main/java/org/dspace/app/itemexport/ItemExportCLITool.java to support `DSpaceRunnable`
   * Add dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportScriptConfiguration.java and dspace-api/src/main/java/org/dspace/app/itemexport/ItemExportScriptConfiguration.java
   * Add `import` and `export` scripts on dspace/config/spring/rest/scripts.xml and remove the rows in dspace/config/launcher.xml
* SAF import and export integration tests:
   * Add ITs dspace-server-webapp/src/test/java/org/dspace/app/itemimport/ItemImportCLIToolIT.java and dspace-server-webapp/src/test/java/org/dspace/app/itemexport/ItemExportCLIToolIT.java
* Minor fixes:
    * Fix the import in case of `metadata_relation.xml` file to build the relationships (before this fix the logic added the metadata without creating the relationship)
    * Fix the import without bitstreams that didn't accept the presence of `collections` or `relationships` files

**Include guidance for how to test or review your PR.**
The ITs can help to test the functionalities.

**OT:**
* I noticed that on https://wiki.lyrasis.org/display/DSDOC7x/Importing+and+Exporting+Items+via+Simple+Archive+Format#ImportingandExportingItemsviaSimpleArchiveFormat-ExportingItems is missing the documentation for the export in zip format (`-z` option).
The export can manage this option as you can see on https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/app/itemexport/ItemExportCLITool.java#L218-L229
* The `collections` file can contain also UUIDs, not only handles, as you can see on https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportServiceImpl.java#L992-L999.
Please update the documentation https://wiki.lyrasis.org/display/DSDOC7x/Importing+and+Exporting+Items+via+Simple+Archive+Format#ImportingandExportingItemsviaSimpleArchiveFormat-DSpaceSimpleArchiveFormat.

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

francescopioscognamiglio committed:
[CST-5699] add scripts for import and export via SAF

francescopioscognamiglio committed:
[CST-5699] add integration tests for import and export via SAF

francescopioscognamiglio committed:
[CST-5699] minor fix to manage collections and relationships files when the contents file is missing

francescopioscognamiglio committed:
[CST-5699] add integration test for relationships functionality

francescopioscognamiglio committed:
[CST-5699] fix to manage relationships by metadata_relation file on SAF import; clean integration tests

francescopioscognamiglio committed:
Merge branch 'main' into CST-5699

francescopioscognamiglio committed:
[CST-5699] minor fix on spurious javadoc params

francescopioscognamiglio committed:
[CST-5699] add integration tests on zip with bitstream and zip with relationships file

francescopioscognamiglio committed:
[CST-5699] Merge branch 'main' into CST-5699

francescopioscognamiglio committed:
[CST-5699] Merge branch 'main' into CST-5699

francescopioscognamiglio committed:
[CST-5699] force to use template to populate dspace.entity.type metadata

mspalti reviewed (approved):
The code looks good! 

I manually tested import, importing from zip, replace and export using a live collection, all from the process UI. Everything worked for me.

I'm glad to see that the ITs cover other scenarios that I didn't test manually. The tests appear comprehensive and passed when I ran them locally.

tdonohue commented on dspace-api/src/main/java/org/dspace/app/itemexport/ItemExportCLITool.java:
```diff
@@ -51,196 +49,175 @@
  * @author Jay Paz
  */
-public class ItemExportCLITool {
+public class ItemExportCLITool extends DSpaceRunnable<ItemExportScriptConfiguration> {
```
In order to allow for different behavior from the CLI and the Admin UI, I think you'll need to split this class into two... there should be an ItemExportCLI class, and an ItemExport class.  See the `org.dspace.,app.bulkedit` classes (`MetadataExportCLI` and `MetadataExport`) as an example

tdonohue commented on dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportCLITool.java:
```diff
@@ -48,348 +47,276 @@
-public class ItemImportCLITool {
-
-    private static boolean template = false;
+public class ItemImportCLITool extends DSpaceRunnable<ItemImportScriptConfiguration> {
```
Same here, you'll need to split this into two classes... likely `ItemImportCLI` and ItemImport`.  See the `org.dspace.app.bulkedit.` classes as examples (`MetadataImport` and `MetadataImportCLI`).

tdonohue commented on dspace-server-webapp/src/test/java/org/dspace/app/itemexport/ItemExportCLIToolIT.java:
```diff
@@ -0,0 +1,342 @@
+ *
+ * @author Francesco Pio Scognamiglio (francescopio.scognamiglio at 4science.com)
+ */
+public class ItemExportCLIToolIT extends AbstractControllerIntegrationTest {
```
These ITs should likely be renamed. They are not testing the CLI version of the tool...instead they test it from the REST API.   If you want a CLI version IT, it should probably be over in `dspace-api`.

tdonohue reviewed (changes requested):
@francescopioscognamiglio : First off, thanks for this set of features!  Overall they work to run the `import` and `export` scripts, but I have a few concerns about the current implementation. I've noticed several bugs and a few security issues:

1. Both the `export` and `import` commands can no longer be run from the CLI.  I've tried running `./dspace export -h` and `./dspace import -h` and neither command works.  I keep getting an error that the command is not found (e.g. `Command not found: import`)

2. For the `export` process, we should *never* allow the `--dest` flag to be specified from the Admin UI. **This is a security issue**, as it potentially allows someone to specify any location on the server where they want their files to be written.   That flag should only be valid from the commandline.  When this is run from the UI, I recommend it be automatically set.  See the `org.dspace.app.bulkedit.MetadataExport` and `MetadataExportCLI` as examples...as the `metadata-export` script already had to implement the same fix.

3. Arguably, when the `export` process is run from the Admin UI, the `--zip` flag should always be enabled & set automatically.   It doesn't make sense to export to a non-Zip file from the Admin UI, as you cannot download an entire directory of exported files.   Additionally when using `--zip`, I noticed that there is currently no way to download the output zip file from the UI, which makes the export tool not useful from the Admin UI.

4. When the `export` process is run from the Admin UI, the `--number` flag should be optional and perhaps default to "1".  Currently it appears to be required, which is frustrating as I have to keep remembering to set it when using the Admin UI.

5. From the Admin UI, the description of the `export` and `import` script is not useful. They just say "Batch Export" and "Batch Import".  I'd recommend we change the description to say "Batch Export to Simple Archive Format (SAF)", and "Batch Import from Simple Archive Format (SAF)"

6.  From the Admin UI, when running the `import` command it should not be possible to specify the `--eperson` flag, as this would allow you to run the script as any other user.  **This is a security issue.**  Instead, the flag should not be displayed and should always default to the current logged in user.  It should still be possible however to use `--eperson` from commandline

7. From the Admin UI, when running the `import` command, the `--zip` param should be required and should allow you to select a file to upload (see how the `metadata-import` script works).  It should also not be possible to specify the `--source` param, as it's not possible to upload an entire directory into the Admin UI (so that `--source` param should only be available from the CLI)

8. From the Admin UI, when running the `import` command,  it appears there no way to obtain the Mapfile?  Maybe I'm overlooking a way to download it.  If that's the case, this makes the `--resume` option impossible to use.

9. BTE no longer exists in DSpace... so the `import` options related to BTE need to be removed. This includes both `--add-bte` and `--inputtype`

Overall, the basic idea is good, and the code is a good start. But, it doesn't entirely work from the Admin UI (e.g. no way to download a Zip file after running `export`... and no way to upload a zip file when running `import`.)

mspalti commented:
Hi @francescopioscognamiglio. I haven't thought much about the differences between existing CLI tasks and the UI processes feature so @tdonohue's feedback is really useful to me as well!  I do recall being able to complete some UI tasks, like import from zip and export. But it's likely that I made these work for me without asking the right questions about how they should work for others, as Tim has done. I'll take a moment to look again.

francescopioscognamiglio commented:
@tdonohue and @mspalti thanks for your feedback, I will work on it.

francescopioscognamiglio committed:
Merge branch 'main' into CST-5699

francescopioscognamiglio commented:
@tdonohue I'm working on it, I will update you asap.

francescopioscognamiglio committed:
Merge branch 'main' into CST-5699

francescopioscognamiglio committed:
[CST-5699] refactoring on SAF import and export functionality

francescopioscognamiglio committed:
Merge branch 'CST-5699' of github.com:4Science/DSpace into CST-5699

francescopioscognamiglio committed:
[CST-5699] minor refactoring

francescopioscognamiglio committed:
[CST-5699] fix SAF import and export functionality; add ITs for SAF via CLI

francescopioscognamiglio commented:
Hi @tdonohue, I implemented all your feedback.
Specifically now we have:
- ItemImport and ItemExport classes: the scripts to be used on the Admin UI side with related ITs (ItemImportIT and ItemExportIT)
- ItemImportCLI and ItemExportCLI classes (extending ItemImport and ItemExport, respectively): the scripts to be used CLI-side with related ITs (ItemImportCLIIT and ItemExportCLIIT)

Please check it out and let me know.
Thank you.

tdonohue reviewed (changes requested):
@francescopioscognamiglio : Thanks for the updates!  My prior concerns seem to have been addressed.  However, I'm still finding a few significant bugs in these scripts when they are executed from the User Interface.  Overall though, they **work**, but there are bugs.

1. A basic `export` works from the `Admin Menu -> Processes -> New`.  However, if I run a second export, it **immediately fails**.  It appears the reason is that the first export's `saf-export.zip` file remains and cannot be overwritten.  
    * This is a problematic design to name every export file the same name.  If two people click export at the same time, it will not work as both exports will attempt to be created at `[dspace]/exports/saf-export.zip`.   I think the name of the file needs to be more unique if it needs to be written to the `[dspace]/exports` folder  (it may need to include the username or the id of the currently logged in session or similar).  Or, maybe there's a way to  ONLY store this `saf-export.zip` as a Bitstream attached to the Process itself?  That'd also work, but it means you cannot write the zip file to the filesystem at any time.
2. The `import` process works. But, every time it runs it logs this: `Deleting temporary zip directory: [dspace]/imports`.  Again, this means that two people cannot perform an import at the same time.  If they do, one will delete the `[dspace]/imports` folder which will cause the other to fail.  This again needs to somehow use a unique directory (perhaps a subpath under `[dspace]/imports`), and we may need to name it based on the user or perhaps the Process ID or similar.

The code overall looks good now, but I have a few inline comments below on things that need cleanup.

francescopioscognamiglio committed:
[CST-5699] delete temp files when added to process; use handler for logging

francescopioscognamiglio commented:
Hi @tdonohue, I implemented the fixes.
Now for the admin UI import and export processes we save the working directory/file in the temporary directory of the file system. We delete the working directory/file when the files are added to the process.
Let me know, thank you.

francescopioscognamiglio committed:
[CST-5699] fix export of a collection with at least two items on SAF export

francescopioscognamiglio commented:
Hi @tdonohue, I fixed a bug in exporting a collection with two or more items. I added the missing ITs for the CLI part to test this case (for the UI part the ITs already exist and were causing the compilation failure).

tdonohue reviewed (changes requested):
@francescopioscognamiglio : I gave this another check, and while it works better, I still think we have a few bugs to work out.

First, I can verify that the previous issues are mostly fixed/resolved.

But, I now see two major new issues (with both of these I was using the Admin UI's Processes page)
1. Each time I run an `export`, I see this in my logs:
    ```
    Exporting Item 635da215-90cc-4a2e-9948-68690d746527, handle 123456789/1120 to /dspace/exports/saf-export.zip/1
    export - 39 @ Attempting to create file /dspace/exports/saf-export.zip/1/metadata_person.xml
    export - 39 @ Attempting to create file /dspace/exports/saf-export.zip/1/dublin_core.xml
    ```
    Notice how the "Attempting to create file.." messages have a non-unique path of  `/dspace/exports/saf-export.zip`? This means that if *two people* run an export at the same time, they will both be writing to that same temporary file of `/dspace/exports/saf-export.zip`.  This will either result in errors, or in unexpected content making it into one person's `saf-export.zip` file.
2. The `import` script is now writing to a temporary location **outside** of DSpace, and it appears to leave its files around.  Here's what I see in my logs:
    ```
    import - 41 @ Set sourceDir using path inside of Zip: /usr/local/tomcat/temp/importSAF335647b6-8a52-4ecb-a8c1-7ebabb199bda4994579856234462436/saf-export.zip12578098849504618650temp/
    import - 41 @ 	Loading dublin core from /usr/local/tomcat/temp/importSAF335647b6-8a52-4ecb-a8c1-7ebabb199bda4994579856234462436/saf-export.zip12578098849504618650temp//1/dublin_core.xml
    ...
    ```
    Notice how the temporary location is now under Tomcat's temp folder (I'd expect it to be using something under `[dspace]/`, e.g. `[dspace]/imports/`).   After the import has completed, if I go to that very long directory under `/usr/local/tomcat`, then I find that all the temporary files are still there.  So, our `import` script is no longer cleaning up after itself. 

Overall, the basic code here looks good, and I appreciate your hard work on this.  But, I do think both of these issues obviously need resolution before we can move this PR forward.  We MUST ensure each process is obviously using a unique folder/file and we also MUST ensure each process cleans up after itself once it is complete.

francescopioscognamiglio committed:
[CST-5699] delete temp dirs after SAF import and export

francescopioscognamiglio commented:
Hi @tdonohue, thanks for your feedback.
I updated the PR with a part to remove temp dirs and files used during import and export functionality.

Let me know any issue.
Have a nice day!

mwoodiupui commented:
Not asking for changes, just an observation in case it is helpful:  I don't see why we have all of these temporary files in the first place, rather than simply adding the data as members of the archive (when exporting) or reading them directly from the archive (when importing).

tdonohue reviewed (changes requested):
@francescopioscognamiglio :  I tested this again today, and it's slightly improved, but I'm still finding bugs:

1. The `import` script is still writing content to a temp directory outside of DSpace (for me it's writing to `/usr/local/tomcat/temp/`.  I'd rather this use the `[dspace]/imports` directory, as that's the temp directory for import processes (and it'd better align it with the `export` script, which already uses `[dspace]/exports`).  I'm worried about having this write to an external temp directory as that's a change in behavior from 6.x & may be unexpected for some sites.
2. The `export` script doesn't appear to support the `--zip` flag when it is run from the Commandline.  I tried running `./dspace export -t COLLECTION -i [uuid] -d [dspace]/temp/ -z export.zip -n 1`   However, the result was that I did not get a Zip file. Instead, Instead, everything is just exported as subdirectories of `[dspace]/temp`.
3. When using the `export` script from the Admin UI, it works. But the Zip file created is "invalid" on Windows (at least I cannot unzip it using Windows-based tools).  It looks to me like the Zip file may be invalid because it includes a double slash before all file paths.  So, the paths look like this when I open the Zip:
    * `saf-export.zip//1/test_pdf.pdf`
    * `saf-export.zip//1/dublin_core.xml`
    * (Notice the double slash "//" before each path, instead of a single slash)

I have been able to round-trip content by exporting (with the -m flag) and then re-importing from the Admin UI. So, the basic scripts do work.  I suspect these last few bugs are more about the temp file location & ability to run the same scripts from the Commandline.

francescopioscognamiglio committed:
[CST-5699] use work dirs instead of temp one

francescopioscognamiglio committed:
[CST-5699] minor fix on zip export from CLI; add missing ITs to check if the exported zip exists

francescopioscognamiglio committed:
[CST-5699] fix duplicate separator on working directories

francescopioscognamiglio committed:
[CST-5699] change test option to validate

francescopioscognamiglio commented:
Thank you for your feedback @tdonohue.

I improved the code:
- both the import and export scripts now write the contents to the [dspace]/imports and [dspace]/exports directories, respectively
- added a small fix on CLI export to zip format and added ITs to check the presence of the exported zip
- fixed duplicate separator on working directories that made the zip "invalid" on Windows

Other improvements:
- to be in line with the other scripts I changed the `-t` option to `-v` to test the import

tdonohue reviewed (changes requested):
@francescopioscognamiglio :  This is now mostly working properly.  Prior feedback has been fixed.  However, I found a few bugs in the new code.  The first is the more severe bug.

* The `import` script leaves files in the `[dspace]/imports/` folder.  If I run that script, and check `[dspace]/imports/`, then I'll find an unzipped copy of the content that was imported.  We obviously need to ensure the import script is cleaning up after itself.
* When I run the `import` script from the commandline, then the import temp directory is *NOT* unique.  For instance, when importing a zip file named `export.zip`, it will use the `[dspace]/imports/export.zip/` folder (and leave the contents there).  I'd expect the commandline script to use a unique location just like when the script is run from the UI.  Since the commandline requires the `--eperson` flag, it could also name the temp directory based on the Eperson UUID (just like it's named with the EPerson UUID when you run this script from the Admin UI).

francescopioscognamiglio committed:
[CST-5699] clean imports dir; use unique location during import via CLI

francescopioscognamiglio commented:
Hi @tdonohue, I fixed the last bugs you found, so now:
- the import (UI and CLI) cleans up the `[dspace]/imports/` folder.
- the import via CLI uses the UUID of EPerson to name the temp directory used for unzip the zip file.

Let me know,
Thank you.

tdonohue reviewed (approved):
👍 Thanks @francescopioscognamiglio !  I've retested this today both from the UI (using https://github.com/DSpace/dspace-angular/pull/1827) and from the commandline.  Everything is working perfectly now & no files are left behind after either an export or import.

So, this now looks great to me.

mspalti reviewed (changes requested):
@francescopioscognamiglio,  @tdonohue  both `import` and `export` worked perfectly for me from the UI and the command line. It's really nice to see this working from the UI! Thanks!

I did notice one issue. The `--help` option doesn't seem to work from the CLI. The error message is: `Missing required options: m, e` These are set up as required options in `ItemImportCLIScriptConfiguration` but it seems like this should be ignored when the `--help` option is chosen. If you provide fake values for `--eperson` and `--mapfile` the help option works, but that's not an ideal solution.

There's sort of a related issue on the Angular side. The `--help` option is listed there but to me it seems unnecessary. Also, if you choose the option you are immediately warned about a missing zip file.

These issues don't affect essential functionality, but it seems like we should try to fix them before release if there's a _quick solution_.  That said, I'd be willing to approve if you think these problems can be addressed later.

tdonohue commented:
@mspalti : I'd recommend we push those to future, minor cleanup as otherwise I have to delay this PR until it is updated again.  Additionally, there are other scripts under `Processes` in the Admin UI which similarly show `--help` as an option (in fact most of them do).  So, it may be more appropriate to have a ticket which is oriented towards  removing `--help` from all scripts when they appear in the Admin UI.

As for the commandline behavior, I noticed that same behavior from `--help`.  But, I think it's good enough as-is.  As you noted the result is not necessarily ideal, but in the end it still shows the help information.

mspalti reviewed (approved):
Happy to approve after your feedback @tdonohue. The help issue is a minor (and almost invisible on the Angular side). It's a bit more problematic for the CLI but I agree it can be fixed later. Overall I'm very glad to see this in 7.4 without delay!
